toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon

```diff
@@ -0,0 +1,157 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon
+
+// --- fail_impl_function_poisoned.carbon
+
+library "[[@TEST_NAME]]";
+
+interface I {
+  fn A(x: Self);
+  fn B();
+}
+
+class B {
+  impl as I {
+    // CHECK:STDERR: fail_impl_function_poisoned.carbon:[[@LINE+3]]:13: error: name `B` used before it was declared [NameUseBeforeDecl]
+    // CHECK:STDERR:     fn A(x: B);
+    // CHECK:STDERR:             ^
+    fn A(x: B);
+    // CHECK:STDERR: fail_impl_function_poisoned.carbon:[[@LINE+11]]:8: note: declared here [NameUseBeforeDeclNote]
+    // CHECK:STDERR:     fn B();
+    // CHECK:STDERR:        ^
+    // CHECK:STDERR:
+    // CHECK:STDERR: fail_impl_function_poisoned.carbon:[[@LINE-9]]:3: error: missing implementation of B in impl of interface I [ImplMissingFunction]
```

Comment:
This second diagnostic doesn't appear to be correct / useful; maybe we should include a TODO to not produce it?

Reply:
Added a TODO.
However, this seems non trivial since we avoid looking into functions with poisoned names so it would be complex to deduce that it was indeed declared in a way that matches the interface function.
The function is technically missing because it's ignored when it's poisoned, but it's indeed very misleading when looking at the diagnostics.
We probably can find some heuristic to avoid triggering this diagnostic when the name is poisoned and attempted to be declared, but I can imagine it would add significant complexity.